enhancements/sig-policy/28-selective-policy-enforcment/README.md

```diff
@@ -0,0 +1,415 @@
+taking effect on a selected subset of the bound clusters.
+
+## Motivation
+
+Selective Policy Enforcement allows fine grained control over the
+timing and application of a policy to clusters. It is intended to
+support this level of control natively within the policy framework. By
+supporting this as a feature users, or higher level
+controllers/orchestrators, do not have to implement a procedural
+pattern of copying policies and manipulating bindings/labels to
+progressivly enforce the policy on clusters.
+
+*Progressive Policy Rollout*
+
+One of the key results of this feature is allowing the impact of a
+Configuration Policy change to be "rolled out" to a fleet of clusters
+in a progressive way rather than all at once. The timing and choice of
+clusters to which the change is applied are typically use case
+dependent but this feature allows scenarios such as
++ Application to a small set of "soak" clusters before rolling the
+  change out to the entire fleet in progressively larger groups
++ Staggered updates to overlapping (geographical or logical) clusters
+  to ensure continuous service coverage
++ Meeting Service Level Agreements for availability by limiting
+  concurrent updates
++ Applying changes to groups of clusters with varying "maintenance
+  windows" when SLAs allow changes to be made
++ Applying changes to one region while deferring other regions
+
+### Goals
+
+The goals of this feature are to
+1. Enable users to select when a policy will be enforced
+1. Enable users to select what subset of bound clusters the policy
+   will be enforced to
+1. Support progressive policy rollouts as described above
+1. Give users visibility into the effect of a Configuration Policy
+   change prior to remediation
+1. Reduce complexity and scalability issues with alternative
+   approaches
+
+### Non-Goals
+
+Out of scope for this feature
+1. A mechanism/feature to capture/define the timing of enforcement
+1. A feature for managing progressive rollouts. The timing and cluster
+   selection are typically use-case specific. There may be enough
+   commonality to define a feature to manage progressive rollouts but
+   that is deferred to a separate discussion/enhancement.
+
+## Proposal
+
+The Selective Policy Enforcement feature adds an API to the policy
+framework to allow enforcement of inform policies to selected
+clusters. The effect of this would be the equivalent of setting the
+complianceType of child (cluster namespace copy) policies to enforce.
+
+The proposed API would be the addition of a remediationAction of
+“controlled” to Policy. In this mode the policy would act as an inform
+policy until the user, controller, or higher level orchestration
+enables remediation directly on the child policy.
+
+The mechanism to enable enforcement of a child policy would be to add
+an annotation to the child policy which enables enforcement
+
+```
+metadata:
+  annotations:
+    versioned-policy-enable: [any|<generation>]
```

Comment:
Instead of an annotation, I think we could do something a little simpler. We could set the `remediationAction` to `controlled` on the parent policy and the replicated policies are created with the `remediationAction` of `inform` as you said.

Then the user would set the `remediationAction` to `enforce` on the subset of replicated policies as desired. The controller (governance-policy-propagator) that manages those replicated policies would ignore changes made to the `remediationAction` on the replicated policies when the parent policy's `remediationAction` is `controlled`.

If the replicated policy gets reevaluated and the spec changes (e.g. parent policy change or Hub policy template resolves differently), then the `remediationAction` gets reset to `inform`.

What do you think?

Reply:
Under the current proposal, I think the `remediationAction` on the replicated policies would be `controlled`, and the framework would infer whether to use `inform` or `enforce` based on the annotation. Is that right @imiller0?

I think the alternative mentioned here would lose out on the `any` value functionality. And I think it might be unclear to the user why the policy reverted to `inform` after they changed it - with the generation annotation and keeping the `remediationAction=controlled`, it might be more clear that something else in the spec has changed, not just that this feature is buggy or something.

Reply:
> Under the current proposal, I think the `remediationAction` on the replicated policies would be `controlled`, and the framework would infer whether to use `inform` or `enforce` based on the annotation. Is that right @imiller0?

That's my understanding too.
> 
> I think the alternative mentioned here would lose out on the `any` value functionality. And I think it might be unclear to the user why the policy reverted to `inform` after they changed it - with the generation annotation and keeping the `remediationAction=controlled`, it might be more clear that something else in the spec has changed, not just that this feature is buggy or something.

That's a good point on the `any` functionality. I think a separate item in the spec of `remediationActionOptions` that could have the keys `informOnUpdate` and `informOnCompliant` (https://github.com/open-cluster-management-io/enhancements/pull/57#discussion_r892339771) would cover these cases.

I just think that an annotation that contains the generation is hard to use because each replicated policy has a different generation. The user also has to be aware of what a generation is which is an obscure Kubernetes concept in my opinion.  It also means that the behavior of resets is different per cluster, which is complex. We also don't have CRD validation on annotations without a validation webhook, which we don't currently have one.

Reply:
I don't think other, more specific remediationAction options would be better. I think it would add complexity at the wrong level. That's personal preference though.

I don't see a basic user trying to use a generation annotation, I think `any` would be more likely if this was being done manually. For a controller adding the annotation, I don't think the generation is an obscure feature, and the fact that it's different on each replicated policy is no problem.

I'm not sure I know what you mean by "the behavior of resets is different per cluster" ... the propagator will always update the spec of the replicated policy to match the root. There's no resetting, unless I misunderstand the proposal. The policy might only be *enforced* at a certain generation, but it will never be out of sync with the root policy.

You bring up a good point with the spoke cluster. I'm now imagining that the spec-sync is where the inform/enforce decision could be made. When it looks at the replicated policy on the hub, and if the remediationAction is `controlled`, then it can look at the annotation & generation and decide whether to set it to `inform` or `enforce`. (A self-managed hub, with no spec-sync, would need extra logic)

Reply:
> I don't think other, more specific remediationAction options would be better. I think it would add complexity at the wrong level. That's personal preference though.

This at least centralizes the complexity in the root policy rather than different behavior per replicated policy.
> 
> I don't see a basic user trying to use a generation annotation, I think `any` would be more likely if this was being done manually. For a controller adding the annotation, I don't think the generation is an obscure feature, and the fact that it's different on each replicated policy is no problem.

If a user is manually using this, I think they want the behavior of enforce until the policy changes. This is the most conservative approach and allows users to opt-in to enforcing on a per-cluster basis. The current proposal in the doc requires you to set the generation in the annotation to get this behavior.
> 
> I'm not sure I know what you mean by "the behavior of resets is different per cluster" ... the propagator will always update the spec of the replicated policy to match the root. There's no resetting, unless I misunderstand the proposal. The policy might only be _enforced_ at a certain generation, but it will never be out of sync with the root policy.

I mean that in the current proposal, a user must set the annotation per replicated policy, which means some can have generation set and some can have `any` set, which is complex to understand.
> 
> You bring up a good point with the spoke cluster. I'm now imagining that the spec-sync is where the inform/enforce decision could be made. When it looks at the replicated policy on the hub, and if the remediationAction is `controlled`, then it can look at the annotation & generation and decide whether to set it to `inform` or `enforce`. (A self-managed hub, with no spec-sync, would need extra logic)

Ideally, a UI or something could just look at the `remediationAction` on the replicated policy on the Hub to know the current state. Calculating the behavior based on the annotation matching the generation of the replicated policy is complex.

Reply:
I like Matt's proposal. We didn't go this way assuming special ignore logic in the sync process was undesired. However, as Matt pointed out, managing different generation values on spokes ties us back to the sync process if we want to achieve consistent behavior.

The main motivation behind this proposal is to have the controlled remediation. The default behavior should be reset on change. We could leave the any case for future and add an annotation if we need to.

Reply:
Would this be a good summary of the proposal?
- remediationAction: [controlled | inform | enforce ]
- controlled will generate child policies with explicit: `remediationAction: inform`
- user/controller/orchestration can set `remediationAction: enforce` on any child policy
- Any change/update to the parent policy resets all child policies to `remediationAction: inform`
- [optional/future] annotation can be added to change the "reset" behavior

Reply:
I believe the main use-case here is for a policy. I want to have `remediation action as inform` for a group of clusters and for the same policy I want to have `remediation action as enforce` for another group of clusters. 

I suggest  it is  better to decouple (or allow override) the remediation Action within the policy definition not by modifying the child policies. To clarify this; As an end user I don’t see (create/update/delete) the child policies I only work with the parent policy. So by adding `remediation Action: controlled` it is not clear how we will let the end user select and modify the child policies that he/she wants to change ?

I believe we better expose the remediation action override API within the policy definition instead of delegating that to the child policies. Like could we have a `clusterSelector` within the policy definition that selects a subset of clusters (from the placement/placementRule’s clusters) And override the `remediation Action` for those clusters subset.

Reply:
> Would this be a good summary of the proposal?
> 
>     * remediationAction: [controlled | inform | enforce ]
> 
>     * controlled will generate child policies with explicit: `remediationAction: inform`
> 
>     * user/controller/orchestration can set `remediationAction: enforce` on any child policy
> 
>     * Any change/update to the parent policy resets all child policies to `remediationAction: inform`
> 
>     * [optional/future] annotation can be added to change the "reset" behavior

Yes, this is exactly what I meant by "settle on an approach that ignores the sync status". My worry is that clusters could potentially get skipped due to the race between an update/sync and an action override, as they are done by two independent players. The user (or TALO) is not aware of the sync status. If there is a sync pending for that child policy, the action override at the child policy level gets reset right away. It may seem like a low runner case, but it makes me rather uncomfortable with this approach.

Initially both Ian and I were trying to avoid another level of grouping/placement/selector kind of approach. Now I feel it seems to be inevitable if consistency across spokes for one rollout process is a strong requirement. I think it's probably a desired (if not mandatory) feature for telco customers at least. Copying policy does give us that.

Reply:
I think the primary use of this feature will be for higher level tools/controllers to control progressive rollouts of changes based on whatever logic/ordering/grouping/timing they define for their use case. In that model enabling one cluster at a time does not seem to be a hard task. With the enabling of enforcement on the child policy there is a range of flexibility in how those higher level tools define logical groups and move clusters into/out of the group. As an example, some real-world rollout strategies would attempt to keep a maximum allowable number of clusters updating at all times. As one cluster completes remediation another one would be enabled (added into the group). These scenarios need the enforcement control to be at a single-cluster granularity (or a flexible grouping construct that allows clusters to be added/removed).

Your point is good though that for a human user to make use of this feature this is a harder interface to manage since each cluster will need to be enabled manually. If there is a way to support both granular levels of control for flexibility as well as group operations that would be good. The current proposal has a bias towards the flexibility and simplicity (eg no additional CRs)

One other consideration, it was intentional in the proposal to avoid any additional "cluster selection" mechanism or secondary set of placement specifications (mentioned in [lines 190-195](https://github.com/open-cluster-management-io/enhancements/pull/57/files#diff-3b6e09ea9bfe2c7568b14f36e957e0792f6af36e5ede839f279d0499b8b02ed1R190-R195)). For example, if users specify "clusterX" in a selector for "enforcement" but that cluster is not included in the placement API the policy won't be enforced and the user will have to debug why.

Reply:
> Would this be a good summary of the proposal?
> 
> * remediationAction: [controlled | inform | enforce ]
> * controlled will generate child policies with explicit: `remediationAction: inform`
> * user/controller/orchestration can set `remediationAction: enforce` on any child policy
> * Any change/update to the parent policy resets all child policies to `remediationAction: inform`
> * [optional/future] annotation can be added to change the "reset" behavior

I like this approach, however, my preference is still as @serngawy stated about using a placement to override the enforcement behavior.

To expand on that, we could have the following new option on the parent policy:

```yaml
remediationAction: inform
remediationActionOverrides:
  enforce:
    - apiGroup: policy.open-cluster-management.io
      kind: PlacementRule
      name: <my placement>
```

For now, we could limit the `remediationActionOverrides` to only allowing the key of `enforce` but the structure would allow us flexibility for other options in the future.

The higher-level controller would still determine which clusters it wants to enforce the policy on if simple label filtering isn't sufficient and it could then create/modify the placement to target clusters by name.

What do you think?

Reply:
> By saying end user I mean any human/controller/tool and I believe in anyway of doing it we will end up having clusterSelector in order to apply the policy remediation for the subset of the bigger policy->placement->clusters group . So how this selection suppose to happen and validated from the policy->placement->clusters space ? may be giving a full use-case example could make this more clear

In my example as shown below, the idea is that your policy would be applied in inform mode against all managed clusters that match the placements bound to the policy using placement bindings like it is today.

Additionally, to select a subset of clusters to enforce, you would specify a `Placement` and/or `PlacementRule` under `remediationActionOverrides.enforce`. If this placement matches clusters that are not in the overall placement of the policy, those clusters would be ignored. Does this make sense?

```yaml
remediationAction: inform
remediationActionOverrides:
  enforce:
    - apiGroup: policy.open-cluster-management.io
      kind: PlacementRule
      name: <my placement>
```

Reply:
yes, make sense @mprahl. its just i didn't saw your comment before i post my comment (we make comment at same time i guess) . What do you think @imiller0 ? I believe this approach will avoid handling the child policy by another tool/controller

Reply:
There are a couple aspects of this override approach that I like but also a couple issues. 

Benefits:
- Using a PlacementRule like this has the advantage of allowing the user/controller/orchestrator to easily select multiple clusters for enforcement. The current proposal requires enabling enforcement on a per-cluster basis. 
- The effect is directly visible in the parent policy. This makes it easier to see what is enabling enforcement.

Issues:
- Duplicate placement APIs. There is a bit of explanation provided under the [implementation details](https://github.com/open-cluster-management-io/enhancements/pull/57/files#diff-3b6e09ea9bfe2c7568b14f36e957e0792f6af36e5ede839f279d0499b8b02ed1R190-R195), but having two sets of placement APIs for the same policy can get confusing/ambiguous when the sets are not completely overlapping. If the inform placement API includes clusters A, B, and C, and the enforce placement APIs include clusters B, E, and F what would be the resulting behavior? The logic would either take the intersection (leaving E and F not enforced) or as distinct rules (enforcing on E and F even though they weren't originally bound to the inform policy). The latter case  would make it very easy to enforce the policy against an unintended cluster when the enforce placement API uses a label that is not a subset of the inform selection.
- Shared ownership of the parent policy. In our use case the parent policy is created through gitops. To enable enforcement the modification by the user/controller/orchestrator would be done directly on the parent policy (not via gitops) leaving the state of the CR shared between the gitops operator and the user/controller/orchestrator. This is very similar issue as occurs when modifying the child policy (in the proposal I summarized above).

Reply:
GitOps can be configured to ignore differences between the source (in git) and the CR on the cluster, but it partly bends the model of gitops (what's in git is reflected in what is on the cluster) and likely increases the amount of work that needs to be done while gitops is reconciling the CRs (deeper inspection and filtering of fields). We could avoid this "2 controllers fighting over the same CR" issue by moving the enforce override into the PlacementBinding rather than the Policy. The user/controller/orchestrator would create both an PlacementBinding (with enforce override) and PlacementRule. This keeps their contribution entirely outside the gitops generated/synced content. This has a couple downsides but perhaps the improved "richness" of the selection syntax makes up for them:
- The user/controller/orchestrator now needs to create, manage, and clean-up these extra CRs. With the approach of manipulating the child policy there aren't any additional CRs needed.
- The issue noted above that when the enforce PlacementRule is not entirely a subset of the inform, the user may be surprised that some clusters are not acted on as they expect. With additional status/logging we should be able to make the reason apparent to a user who goes looking
- The enforce action is decoupled from the parent policy. Anyone looking for why a child policy is being enforced would need to be aware of an override that exists solely within the PlacementBinding. Can the status field of the policy be updated to include information on an override?

In terms of implementation, would a second PlacementBinding result in an additional child policy? For scaling reasons this would be an issue.

From the perspective of a controller that would use this, I'm curious what @jc-rh thinks from prior work in this area?

Something like:
```yaml
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
...
placementRef:
  apiGroup: apps.open-cluster-management.io
  kind: PlacementRule
  name: override-placementrule
subjects:
- apiGroup: policy.open-cluster-management.io
  kind: Policy
  name: test-policy-1
  remediationActionOverride: [enforce|inform]
```

Reply:
Just realized that the dependecy between the rules should be inverted:
```
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
  name: enforce-placementrule1
  namespace: consumer-controller-managed-ns
spec:
  baseRuleRef:
    name: inform-placementrule1
    namespace: gitops-managed-ns
  clusterSelector: 
...
```
We would still need to create a new binding. It would be great if the binding can specify subjects from a different namespace.

Reply:
Just realized that the dependecy between the rules could be inverted:
```
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
  name: enforce-placementbinding1
  namespace: consumer-controller-managed-ns
spec:
  placementRef:
    apiGroup: apps.open-cluster-management.io
    kind: PlacementRule
    name: enforce-placementrule1
    basePlacementRef:
      name: inform-placementrule
      namespace: gitops-managed-ns
  subjects:
  - apiGroup: policy.open-cluster-management.io
    kind: Policy
    name: test-policy-1
    namespace: gitops-managed-ns
...
```
The new binding has a ref to the base rule in addition to the override rule. This way there won't be share ownership on the binding. But there is no guarantee on whether the base rule is indeed associated with the policy ...

Reply:
@jc-rh that's what I was thinking. Could you just have a placement rule override per policy to avoid conflicts if it's too complex for TALO to sync it across all policies?

Reply:
I was assuming one override rule per policy. The sharing case is where multiple CGUs remediating the same policy on different set of clusters. With this approach, TALO will have to update the cluster list in this shared rule without overwriting the clusters from other CGU. That's why I had multiple rules in my override through binding example.

Single rule vs multiple rule aside, I feel that binding is better holder for the override rule than policy for a couple of reasons

- It's consistent with the general placement design
- It allows us to link the override rule with the base rule so that it's clear the override rule acts as a subfilter on top of a particular base rule (consider a policy can be placed through multiple bindings/rules)

However, I am not sure if there is reliable way of finding the binding that does the placement of the policy on a particular cluster ...

Reply:
@jc-rh  I think it will be much simpler if you consider moving the managedCluster in/out from the `remediationActionOverrides` placement/placementRule. For example; let's assume the policy placmentRule has a clusterSelector  as below

`  clusterSelector: 
  matchExpressions: 
  - key: common
    operator: In
    values:
    - "true"
`
AND for the `remediationActionOverrides` placement/placementRule has a clusterSelector as below

`  clusterSelector: 
  matchExpressions: 
  - key: enforce
    operator: In
    values:
    - "policyName"
`
So what TALO can do is update the ManagedCluster CR labels to include the matching label `enforce: policyName`  then the cluster will be part of `remediationActionOverrides` placementRule clusters list. once the policy becomes compliant TALO remove/update the label `enforce: policyName`  from the ManagedCluster CR.

Reply:
I think we just completed a full circle. We said it might be better to have a group thing so we don't label child polices one by one. Now you are saying we label managed cluster one by one :)

Reply:
Where did we end up on this discussion?

I'm thinking that I don't want to support a user/operator modifying the replicated policy's remediation action. I want to keep that replicated policy as an implementation detail and not as a new interface for users to further alter policy behavior.

My preference is to allow a placement binding to override the remediation action of a subset of clusters that the overall placement selects.

@JustinKuli what do you think?

Reply:
> To be honest I got a little lost in this discussion... we might need a summary of the current options. Sorry sweat_smile

I don't blame you haha.

I'll try to summarize.

## Option 1

Specify an additional placement in the parent policy of the subset of clusters that should have the remediation action set to enforce.

```yaml
remediationAction: inform
remediationActionOverrides:
  enforce:
    - apiGroup: policy.open-cluster-management.io
      kind: PlacementRule
      name: <my placement>
```

## Option 2

Specify a remediation action override for a subset of clusters in the `PlacementBinding`.

```yaml
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
...
spec:
  placementRef:
    apiGroup: apps.open-cluster-management.io
    kind: PlacementRule
    name: inform-placementrule
    overridePlacementRef:
      - apiGroup: apps.open-cluster-management.io
        kind: PlacementRule
        name: enforce-placementrule1
      - apiGroup: apps.open-cluster-management.io
        kind: PlacementRule
        name: enforce-placementrule2      
  subjects:
  - apiGroup: policy.open-cluster-management.io
    kind: Policy
    name: test-policy-1
```

## Option 3

This is the proposal in the doc in the PR. This proposes setting the parent policy remediation action to `controlled` and allow users and other processes to set the `remediationAction` in each replicated policy.

Reply:
Call me old school, but I like the original proposal - option 3.

I don't mind that a user/controller would be able to modify a field on the replicated policies. It feels like the proposal has considered how to do this safely, and if a rollback is needed it is very easy to remove the `controlled` bit from the parent policy. The main advantage I see is that as a user, I can just look at the Policy objects on the hub and understand what is going on.

Reply:
Resolving this thread based on the discussion and subsequent updated proposal.